gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java

```diff
@@ -288,6 +303,15 @@ private static void modifyStateIfRetryRequired(org.apache.gobblin.configuration.
     state.removeProp(TimingEvent.FlowEventConstants.DOES_CANCELED_FLOW_MERIT_RETRY);
   }
 
+  private static boolean isStateTransitionToFinal(org.apache.gobblin.configuration.State currentState, List<org.apache.gobblin.configuration.State> prevStates) {
+    Set<String> finalStates = ImmutableSet.of(ExecutionStatus.COMPLETE.name(), ExecutionStatus.CANCELLED.name(), ExecutionStatus.FAILED.name());
+    if (prevStates.size() == 0) {
+      return finalStates.contains(currentState.getProp(JobStatusRetriever.EVENT_NAME_FIELD));
+    }
+    return currentState.contains(JobStatusRetriever.EVENT_NAME_FIELD) && finalStates.contains(currentState.getProp(JobStatusRetriever.EVENT_NAME_FIELD))
+        && !finalStates.contains(prevStates.get(prevStates.size()-1).getProp(JobStatusRetriever.EVENT_NAME_FIELD));
```

Comment:
thought marginally more processing, it more clearly captures the intent and handles a corner case (if even possible) of the state flip flopping from final to non-final back to final:
```
Stream<Boolean> statesFinality =
    Stream.concat(prevStates.stream(), Stream.of(currentState)).map(isFinalState);
return statesFinality.filter(identity).count() == 1 && Iterables.getLast(statesFinality) == true;
```

Reply:
I think it's impossible for a state to flip flop because of the way how we perform `mergeState`, which follows an ordered execution status of 
```
  private static final List<ExecutionStatus> ORDERED_EXECUTION_STATUSES = ImmutableList
      .of(ExecutionStatus.COMPILED, ExecutionStatus.PENDING, ExecutionStatus.PENDING_RESUME, ExecutionStatus.PENDING_RETRY,
          ExecutionStatus.ORCHESTRATED, ExecutionStatus.RUNNING, ExecutionStatus.COMPLETE,
          ExecutionStatus.FAILED, ExecutionStatus.CANCELLED);
```
I think there is a risk if there is a COMPLETE event followed by a FAILED or CANCELLED event but I believe the first "final" event state we run into is the one we want users to accept.

Reply:
agreed... that's what I meant by perhaps not possible.  so the difference would be purely stylistic... either form looks fine